Reordering the current Stacks reference docs [DO NO MERGE] (#698)
hashicorp/web-unified-docs: content/terraform/v1.12.x/docs/language/block/stack/deploy/deployment.mdx

---
page_title: deployment block reference for Terraform Stack deployments
description: Use the `deployment` block to define how many times you want to deploy your Stack's infrastructure.
---

# `deployment` block reference

Use the `deployment` block to define how many times you want to deploy your Stack's infrastructure. You can define `deployment` blocks in a `<NAME>.tfdeploy.hcl` file, which is the deployment configuration file for your Stack.

## Background

<Note>

HCP Terraform supports up to a maximum of 20 deployments per Stack.

</Note>

> The `deployment` block defines how to deploy your Stack's infrastructure. Each deployment represents a separate instance of your Stack infrastructure. Each deployment runs in its own isolated agent, wholly separated from other Stack deployments.

Reviewer: Hmm, "isolated agent" isn't quite right. I mean, every plan that runs has some isolation, but we don't lock a deployment to a specific agent. 

```suggestion
The `deployment` block defines how to deploy your Stack's infrastructure. Each deployment represents a separate instance of your Stack infrastructure. Each deployment maintains its own isolated state data, wholly separated from other Stack deployments.
```

I think that's probably closer to what's meant?

Each Stack requires at least one `deployment` block, and you can add additional `deployment` blocks every time you want to deploy your Stack's infrastructure again with different input values. To learn more, refer to [Define deployment configuration](/terraform/language/stacks/deploy/config).

> You can define the following configuration blocks in a `tfdeploy.hcl` file:

Reviewer: ```suggestion
In addition to the `deployment` block, you can define the following configuration blocks in a `tfdeploy.hcl` file:
```

> - Use the [`deployment_group`](/terraform/language/block/stack/deploy/deployment-auto-group) block to create a group of deployments and apply rules to those deployments.
> - Use the [`deployment_auto_approve`](/terraform/language/block/stack/deploy/deployment-auto-group) block to define rules that automatically approve deployment plans based on specific conditions.

Reviewer: ```suggestion
- Use the [`deployment_auto_approve`](/terraform/language/block/stack/deploy/deployment-auto-group) block to define rules that automatically approve deployment plans based on specific conditions.
- Use the [`deployment_group`](/terraform/language/block/stack/deploy/deployment-auto-group) block to assign `deployment_auto_approve` rules to specific deployments.
```

Until we get the cooler group stuff running, I think that's an easier way for a user to sort this out in their head? It does mean this description needs an update once we do build all that, but it seems likely to need that anyway.

- Use the [`identity_token`](/terraform/language/block/stack/deploy/identity_token) block to authenticate a deployment with a provider configuration using OIDC.
- Use the [`store`](/terraform/language/block/stack/deploy/store) block to define key-value secrets in your deployment configuration and use existing variable sets in HCP Terraform.
- Use the [`locals`](/terraform/language/block/stack/deploy/locals) block to define local values that you can reuse within your deployment configuration.
- Use the [`publish_output`](/terraform/language/block/stack/deploy/publish_output) block to output values that you want to be available to other Stacks.
- Use the [`upstream_input`](/terraform/language/block/stack/deploy/upstream_input) block to source inputs from another Stack.

## Configuration model

The `deployment` block supports the following arguments:

- [`deployment "<LABEL>"`](#deployment) &nbsp; block
  - [`inputs`](#inputs) &nbsp; map
  - [`deployment_group`](#deployment_group) &nbsp; reference to deployment group

>   - [`destroy`](#destroy) &nbsp; boolean

Reviewer: ```suggestion
  - [`destroy`](#destroy) &nbsp; boolean
  - [`import`](#TODO) boolean
```

That `import` argument is used for triggering import runs, and is used by the migrate-from-workspace tool.

Author: Ooooh I had no idea about this - thank you SO MUCH for calling this out! I'm adding this as a todo and I'll find out who to talk to on the migrate team.

## Complete configuration

All available arguments are defined in the following `deployment` block:

```hcl
deployment "<LABEL>" {
  inputs = {
    <VARIABLE_NAME> = <VALUE>
  }
  deployment_group = "deployment_group.<NAME>"
  destroy = <BOOLEAN>
}
```

## Specification

A `deployment` block supports the following configuration.

### `deployment "<LABEL>"`

> The label after the `deployment` keyword is a name for the deployment, which must be unique among all deployments in the same deployment configuration file. The name of a deployment can be any valid [identifier](/terraform/language/syntax/configuration#identifiers).

Reviewer: This might be a good place to mention that you can't really rename a deployment — changing the label results in a destroy of the old deployment name and a create of the new deployment name.

The `deployment` block supports the following arguments:

| Argument | Description | Type | Required? |
| --- | --- | --- | --- |
| `inputs` | A mapping of Stack variable names to values for this deployment. | Map | Required |
| `deployment_group` | A reference to the group that this deployment belongs to. | Reference | Optional |
| `destroy` | A boolean flag that indicates whether HCP Terraform should destroy this deployment. | Boolean | Optional |

### `inputs`

The `inputs` argument defines a mapping of Stack variable names to values for this specific deployment.

```hcl
deployment "<LABEL>" {
  inputs = {
    <VARIABLE_NAME> = <VALUE>
    <VARIABLE_NAME_TWO> = <VALUE_TWO>
  }
}
```

> The keys in the `inputs` map must correspond to the names of the variables defined in your component configuration, and the values must be valid HCL literals that meet the type constraints of those variables.

Reviewer: Doesn't have to be literals, you can use expressions here as well! That's how you're able to reference locals, identity tokens, and values from store objects. 

```suggestion
The keys in the `inputs` map must correspond to the names of the variables defined in your component configuration, and the values must be valid HCL expressions that meet the type constraints of those variables. Input value expressions can reference `locals`, `identity_token`, `store`, and `upstream_value` blocks. 
```

#### Summary

> - Data type: Map

Reviewer: idk if we're making this distinction rigorously in the docs these days, but it's an object, not a map. 

- map: key/value where all values are the same _type_.
- object: key/value where values can be different types. 

To a user working directly in the terraform language there's not really any practical difference.......... it's only critical once you're in a provider implementation or sourcing data from weird places.

Reviewer: anyway, depending on current practice elsewhere in language docs, feel free to disregard this

Author: Oh dang, thank you so much! I didn't actually know this! I'll make this object  to match and will actually make sure we align other places accross the board!
